src/awkward/operations/ak_from_raggedtensor.py

```diff
@@ -79,9 +63,18 @@ def _impl(array):
 
 
 def _tensor_to_np_or_cp(array, device):
-    import tensorflow as tf
+    if device.endswith("GPU", 0, -2):
```

Comment:
I had to check the documentation on [str.endswith](https://docs.python.org/3/library/stdtypes.html#str.endswith), but it seems that this is equivalent to

```python
    if device[:-2].endswith("GPU"):
```

(though I think the latter is easier to understand because slicing is more well-known than the extra arguments of `str.endswith`).

However, are you assuming that the GPU number is one digit? That is, will the above code break for a computer with 10 GPUs?

If the format for the 15th GPU is `"GPU-14"` (zero-indexed), then maybe you want

```suggestion
    if device.split("-")[0] == "GPU":
```

(and if lowercase is possible, you can also add a `.upper()` in the chain).

But before you accept the suggestion above, is it really a hyphen? If there's only one GPU, would there be no hyphen? (Note that `device.split("-")[0]` is equal to `device` if there is no hyphen, so the same code may be fine.)

Reply:
If there's only one GPU, then the `device` looks like this: `/job:localhost/replica:0/task:0/device:GPU:0`
So, I think `if device.split(":")[-2] == "GPU":` will work.

Reply:
It will, but it relies on TensorFlow never changing the text to end with `"GPU"` rather than `"GPU:0"`. All of this is about trying to write something [defensively](https://en.wikipedia.org/wiki/Defensive_programming), so that either our incomplete knowledge of the upstream library (TensorFlow) or possible changes in that upstream library would cause our code to break. By "break," I mean "do the wrong thing without an error message." Failing with an error message if TensorFlow changes would be fine.

Given that what we expect from TensorFlow is a string like

```
/job:localhost/replica:0/task:0/device:CPU:0
```

or

```
/job:localhost/replica:0/task:0/device:GPU:0
```

or

```
/job:localhost/replica:0/task:0/device:GPU:14
```

this would be a safe way to catch it:

```python
import re

m = re.match(".*:(CPU|GPU):[0-9]+", device)
if m is not None:
    raise NotImplementedError(f"TensorFlow device has an unexpected format: {device!r}")
if m.groups()[0] == "GPU":
    ...
```

It also expresses to the future maintainer (or code reviewer) what you know about what TensorFlow gives you. (The import needs to be in the import section.)

Reply:
thank you!